Add encoder calibration documentation

anirudhupadhyaya opened the pull request.

anirudhupadhyaya committed:
initial commit

anirudhupadhyaya committed:
Add content

anirudhupadhyaya committed:
Update docs

npetersen2 commented:
@anirudhupadhyaya requests that I review this PR in terms of direction of content

npetersen2 commented:
@anirudhupadhyaya this is on the right track and seems to be written to the right audience, i.e., a user of the AMDC who is trying to interface an encoder to their control system.

There are a few typos and awkward sentences which should be resolved.

However, my main feedback would be to add a "master" block diagram which shows all the possible steps in processing the encoder integer feedback value into something useful for the control system. Then, replace your C code blocks with a *single* block of C code which implements the block diagram for "all cases" i.e., CW or CCW, etc.

I am not sure about your `while()` loop for performing the `mod(x, 2*pi)` operation. Probably better to just use `mod()` here unless you have a good reason not to use it.

Also, add a section about validating the encoder offset via the d-axis voltage during operation.

I am also curious to see your section on converting position to speed.

anirudhupadhyaya committed:
Update doc

anirudhupadhyaya committed:
Update docs

anirudhupadhyaya committed:
Update docs

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -2,85 +2,96 @@
 ## Calibration
 
-Incremental encoders which are typically used with AMDC have a fixed number of counts per revolution (for example, 1024) and is denoted by `CPR`. The user needs needs to convert the count into usable angular information which maybe used in the code.
+Incremental encoders which are typically used with AMDC have a fixed number of counts per revolution (for example, 1024) and is denoted by `CPR`. The user needs needs to write some code to obtaine the encoder count at a given instance and tconvert the count into usable angular information which can be used within the control code.
```
typo

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -2,85 +2,96 @@
+<img src="./resources/EncoderCodeBlockDiargam.svg" width="100%" align="center"/>
 
-Example code for when encoder count is increasing with CCW rotation of shaft:
+As a first step, the user may use the AMDC `enc` driver function `encoder_get_position()` to get the count of the encoder reading. Next, the user needs to verify if the encoder count is increasing or decreasing with counter-clock wise rotation of shaft. This may be done by manually rotating the shaft and observing the trend of the reported position with respect to the direction of rotation. This  ...
```
Replace "`enc` driver function" with "`drv/encoder` driver module"

npetersen2 replied:
Add commentary about the difference between the `steps` and `position` output from the driver

anirudhupadhyaya replied:
@npetersen2 could you point me to some material on the difference between `steps` and `position`? I am not clear on this add an commentary

elsevers replied:
`steps` indicates incremental movement, whereas `position` is the actual shaft angle/position

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -2,85 +2,96 @@
+### Finding the offset
 
-    enc_theta_m_offset = 100;
+The example code shown above makes uses of an encoder offset value. It is necessary for the user to find this offset experimentally for their machine. For synchronous machines, this offset is the count value measured by the encoder when the d-axis of the rotor is aligned with the phase U winding of the stator. 
```
"makes use"

npetersen2 replied:
"phase U winding axis"

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -2,85 +2,96 @@
 
-    // Convert to radians
-    theta_m_enc = (double) PI2 * ( ( (double)ENCODER_COUNT_PER_REV - (double)1 -(double)position + enc_theta_m_offset )/ (double) ENCODER_COUNT_PER_REV);
+To get the rotor to align with the phase U winding, the user may have some current flow through phase U and out of phase V and phase W. This can done by using a DC supply, with phase U connected to the positive terminal and phases V and W connected to the negative terminal. Alternately, the user may also inject some current along the d-axis using the AMDC injection function, but with the rotor po ...
```
Add more commentary about "AMDC injection function"....a user must have already set this up in your control code...

npetersen2 replied:
How does someone "override the rotor position to 0" ?

npetersen2 replied:
"appropirate" typo

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -2,85 +2,96 @@
+$$
+
+
+$\Omega_\text{raw}$ will be a choppy and noisy signal due to the derivative operation. A low pass filter may be applied to this signal as shown below to get a filtered speed, $\Omega_\text{lpf}$. The user may select an appropriate bandwidth, $\omega_b$ for the low pass filter to eliminate the noise introduced by the derivative operation. However, this signal will always be a lagging estimate of t ...
```
"due to the derivative operation and the digital nature of the incremental encoder"

npetersen2 replied:
Can you provide some guidance on selecting the filter bandwidth? What would be a good first choice for someone to use?

npetersen2 commented on source/getting-started/control-with-amdc/encoder-fb/resources/EncoderCodeBlockDiargam.svg:
Thanks for adding this image, but this is not what I envisioned.

Can you change this to a block diagram which is like what you would find in simulink? i.e., a signal flows from left to right and is manipulated? This block diagram should match your C code implementation, just in a graphical form.

anirudhupadhyaya committed:
Update doc

anirudhupadhyaya commented:
I am yet to update the block diagram, but I have addressed rest of the comments.

anirudhupadhyaya committed:
Update block diagram

anirudhupadhyaya commented:
Block diagram updated.

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
 
+## Background
+
+Encoders provide rotor position feedback to the control system in a motor drive. This document describes a method to convert the raw readings of an encoder into meaningful position information which can be used by the contrl algorithm. Consequently, few techniques to derive the rotor speed from the measured position are also proposed. For more information on how an encoder works and how they may  ...
```
> Consequently, few techniques to derive the rotor speed from the measured position are also proposed.

This is a confusing sentence. It could mean that only a few techniques exist in the world. Or that we are proposing techniques here for the first time ever. 

Do you need the sentence -- are you trying to convey information that is not already in the previous sentence?

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
-- Converting from raw counts to "theta"
-- Direction: +/- depending on phase connections
-- Sync machines: dq offset
+Incremental encoders which are typically used with AMDC have a fixed number of counts per revolution (for example, 1024) and is denoted by `CPR`. The user needs needs to write some code to obtain the encoder count at a given instance and convert the count into usable angular information which can be used within the control code.
```
```suggestion
Incremental encoders are typically used with the AMDC and have a fixed number of counts per revolution `CPR` (for example, `CPR = 1024`). The user needs needs to provide code that interfaces to the AMDC's drivers to read the encoder count and convert it into usable angular information that is suitable for use within the control code.
```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+
+<img src="./resources/EncoderCodeBlockDiargam.svg" width="100%" align="center"/>
+
+As a first step, the user may use the AMDC `drv/encoder` driver module `encoder_get_position()` to get the count of the encoder reading. The`drv/encoder` driver module also has a function called `encoder_get_steps()` which gives the incremental change in the encoder position. Whereas, `encoder_get_position()` gives the actual position of the shaft and this can be converted the rotor position in r ...
```
I think you mean for this paragraph to be describing your figure? Please tell us that and place the paragraph *above* your figure. Start with something like "The recommended approach to reading the shaft position from the encoder is illustrated in the figure below."

Please break this up into multiple paragraphs, with each paragraph corresponding to one of the blocks in the figure:
- First paragraph is discussing obtaining the encoder reading
- Start a new paragraph at "Next, the user needs..." (to discuss direction of rotation)
- Final paragraph is discussing wrapping the encoder value with the `mod` function.

Some improvements in your figure labeling and the text of these paragraphs may be needed to help this make sense. For example, you are referring to the "offset" here, but aren't telling us what that means. Need to explain.

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+
+### Finding the offset
+
+The example code shown above makes uses of an encoder offset value. It is necessary for the user to find this offset experimentally for their machine. For synchronous machines, this offset is the count value measured by the encoder when the d-axis of the rotor is aligned with the phase U winding axis of the stator. 
```
Based on how you introduce offset above (see my comment above) you might need to slightly tweak this text

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+
+The example code shown above makes uses of an encoder offset value. It is necessary for the user to find this offset experimentally for their machine. For synchronous machines, this offset is the count value measured by the encoder when the d-axis of the rotor is aligned with the phase U winding axis of the stator. 
+
+To get the rotor to align with the phase U winding, the user may have some current flow through phase U and out of phase V and phase W. This can done by using a DC supply, with phase U connected to the positive terminal and phases V and W connected to the negative terminal. Alternately, the user may also inject some current along the d-axis using the AMDC [Signal Injection](https://docs.amdc.dev/ ...
```
Can you make this a bit easier to read? This somehow reads complicated when it is actually quite a simple procedure. Say something like: 

> To determine the appropriate offset value (reported as `enc_theta_m_offset` in the code example above), eliminate any source of load torque on the shaft and apply a large current vector at 0 degrees ($I_u = I_0$, $I_v = I_w = -\frac{1}{2} I_0$). This should cause the rotor to align with the phase u axis. The offset value can now be obtained as `enc_theta_m_offset = encoder_get_position()`.

(Do I have this right? Please edit / fix as appropriate)

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
-- State Filter
-- Observer
\ No newline at end of file
+The user needs to compute a rotor speed signal from the obtained position signal to be used in the control algorithm. There are several ways to do this. A simple and straightforward way to do this would be to compute the discrete time derivative of the position signal in the controller as shown below. This can be referred to as $\Omega_{raw}$.
```
```suggestion
The user needs to compute a rotor speed signal from the obtained position signal to be used in the control algorithm. There are several ways to do this. 

### Difference Equation Approach

A simple, but naive, way to do this would be to compute the discrete time derivative of the position signal in the controller as shown below. This can be referred to as $\Omega_{raw}$.
```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+$$
+
+
+$\Omega_\text{raw}$ will be a choppy and noisy signal due to the derivative operation and the digital nature of the incremental encoder. A low pass filter may be applied to this signal as shown below to get a filtered speed, $\Omega_\text{lpf}$. The user may select an appropriate bandwidth, $\omega_b$ for the low pass filter to eliminate the noise introduced by the derivative operation. A recomme ...
```
```suggestion
Unfortunately, using this approach results in noise in $\Omega_\text{raw}$ due to the derivative operation and the digital nature of the incremental encoder. 

### Low Pass Filter Approach

To solve this, _a low pass filter_ may be applied to this signal. This is shown below to obtain a filtered speed, $\Omega_\text{lpf}$.
```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+ \Omega_\text{lpf}[k] =  \Omega_\text{raw}[k](1 - e^{\omega_b T_s}) + \Omega_\text{lpf}[k-1]e^{\omega_b T_s}
+$$
+
+An observer which implements a mechanical model of the rotor as shown below will produce a no-lag estimate of the rotor speed, denoted by $\Omega_\text{sf}$. To implement an observer, the user needs to know the system parameters - `J` - the inertia of the rotor shaft and `b` - the damping coefficient of the rotor shaft. Further, to obtain a no-lag estimate it is necessary to provide the electrome ...
```
```suggestion
### Observer Approach

To obtain a no-lag estimate of the rotor speed, users may create an observer which implements a mechanical model of the rotor as shown below. 

<img src="./resources/ObserverFigure.svg" width="100%" align="center"/>

```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+
+$$
+K_p = \omega_{sf}b, K_i = \omega_{sf}J
+$$
```
The will produce a no-lag estimate of the rotor speed, denoted by $\Omega_\text{sf}$. To implement an observer, the user needs to know the system parameters - `J` - the inertia of the rotor shaft and `b` - the damping coefficient of the rotor shaft. Further, to obtain a no-lag estimate it is necessary to provide the electromechanical torque, $T_{em}$ as input to the mechanical model. The `P-I` part of the observer closes the loop on the speed with $\Omega_\text{raw}$ being the reference input. The recommended tuning approach is as follows:

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+$$
+
+An observer which implements a mechanical model of the rotor as shown below will produce a no-lag estimate of the rotor speed, denoted by $\Omega_\text{sf}$. To implement an observer, the user needs to know the system parameters - `J` - the inertia of the rotor shaft and `b` - the damping coefficient of the rotor shaft. Further, to obtain a no-lag estimate it is necessary to provide the electrome ...
+
```
```suggestion

The estimate of rotor speed is denoted by $\Omega_\text{sf}$. To implement this observer, the user needs to know the system parameters:
- `J`: the inertia of the rotor  
- `b` the damping coefficient of the rotor. 

It is also necessary to provide the electromechanical torque, $T_{em}$ as input to the mechanical model. 

The `PI` portion of the observer closes the loop on the speed, with $\Omega_\text{raw}$ being the reference input. The recommended tuning approach is as follows:
```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+K_p = \omega_{sf}b, K_i = \omega_{sf}J
+$$
+
+This tuning ensures a pole zero cancellation in the closed transfer function, resulting in a unity transfer function for speed tracking under ideal parameter estimates of `J` and `b`.  A recommended bandwidth for the observer is 10 Hz.
```
```suggestion
This tuning ensures a pole zero cancellation in the closed transfer function, resulting in a unity transfer function for speed tracking under ideal parameter estimates of `J` and `b`.  An observer bandwidth of 10 Hz is typical of most systems, but similar to the low pass filter approach, users may need to alter this based on the unique aspects of their system.
```

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,97 @@
+
+This tuning ensures a pole zero cancellation in the closed transfer function, resulting in a unity transfer function for speed tracking under ideal parameter estimates of `J` and `b`.  A recommended bandwidth for the observer is 10 Hz.
+
+<img src="./resources/ObserverFigure.svg" width="100%" align="center"/>
```
```suggestion

```

elsevers reviewed (changes requested):
Thanks @anirudhupadhyaya. I think this is on the right track. I have left you a few improvement requests below, and have gone ahead and implemented most of the needed changes as suggestions that you can simply accept. 

In addition to this, I would like to see you add one (or more) reference papers to help users implement and think about the observer approach. There must be some good material from our digital control study group that Nathan and I did with you and Aravind that you can cite here?

You'll get big bonus points from me if you can include a MATLAB / Simulink example of a motor with an observer. I'd really love it if you can do this, but maybe too late in the game for this though.

anirudhupadhyaya committed:
Apply suggestions from code review

Co-authored-by: Eric Severson <eric.severson@wisc.edu>

anirudhupadhyaya committed:
Update doc to address review comments

anirudhupadhyaya commented:
@elsevers I have addressed your review comments.

elsevers reviewed (changes requested):
Hmm, something seems to have gone wrong here. 

<img width="1043" alt="image" src="https://github.com/Severson-Group/docs.amdc.dev/assets/7421348/cda98166-8e04-4083-b860-76d8467ce146">

We're missing the discussion on how the low pass filter always producing a lagging estimate. I remember suggesting changes to this --> how 10 Hz is a good starting bandwidth, that it will always produce a lagging estimate, etc. 

I can't seem to find the text I had edited for this. Do you think you accidentally rejected my suggestion?

anirudhupadhyaya commented:
Oh.. I guess it went away when I accepted the suggestions?

anirudhupadhyaya committed:
Add line on low pass filter

anirudhupadhyaya commented:
@elsevers I added a line on the bandwidth recommendation for low pass filter.

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,130 @@
+ \Omega_\text{lpf}[k] =  \Omega_\text{raw}[k](1 - e^{\omega_b T_s}) + \Omega_\text{lpf}[k-1]e^{\omega_b T_s}
+$$
+
+10 Hz is a good bandwidth for the low-pass filter to start with, but users may need to alter this based on the unique aspects of their system. However, this approach will always produce a lagging speed estimate.
```
```suggestion
Here, $T_{\rm s}$ is the control sample rate and $\omega_b$ is the low pass filter bandwidth. The user must select this bandwidth to obtain a sufficiently clean speed signal.  The optimal bandwidth to use is going to vary based on the motor system. Typically, a bandwidth of 10 Hz is a reasonable starting point. This can be reduced if the speed signal remains too noisy, or increased for higher speed controls. 

Note that this low pass filter approach will always produce a lagging speed estimate due to phase delay in the filter transfer function. This may be unacceptable higher performance motor control algorithms.
```

anirudhupadhyaya committed:
Update source/getting-started/control-with-amdc/encoder-fb/index.md

Co-authored-by: Eric Severson <eric.severson@wisc.edu>

npetersen2 commented:
# Simulink PLL Block for Encoder Feedback Filtering

![image](https://github.com/Severson-Group/docs.amdc.dev/assets/20168990/6f85b13a-feab-4f59-8267-896af489bf61)

![image](https://github.com/Severson-Group/docs.amdc.dev/assets/20168990/c174560e-de7a-4b08-b3b1-cdf60b9912ad)

Zip file of the `.slx` file for Simulink:
 
[pll_encoder.zip](https://github.com/user-attachments/files/16057338/pll_encoder.zip)

elsevers committed:
Edit background

elsevers committed:
Fix file name type-o

elsevers committed:
Fix type-os

elsevers committed:
Add config instructions

noguchi-takahiro commented:
@elsevers to my understanding, we should not have -1 in the second equation, i.e., 

- if CCW = 1, 
$\theta_\text{CCW} =2\pi \frac{\text{position} - \text{offset}}{\text{ENCODER COUNT PER REV}}$
- else if CCW = 0
$\theta_\text{CW} =2\pi \frac{\text{ENCODER COUNT PER REV} - \text{position} + \text{offset}}{\text{ENCODER COUNT PER REV}} = 2\pi \frac{\text{ENCODER COUNT PER REV}}{\text{ENCODER COUNT PER REV}}-2\pi \frac{\text{position}-\text{offset}}{\text{ENCODER COUNT PER REV}}=2\pi-\theta_\text{CCW}$

Your proposed changes works too.

I noticed that this exists from ARL codebase to operate the BP4: 
https://github.com/Severson-Group/ARL-eturbo/blob/c1cfcf5c685953966a533b32e2c0184f8654331f/Embedded/My-C-Code/usr/bm_4dof/task_bm_4dof.c#L187

elsevers committed:
Edit first 1/3 of document

elsevers commented:
Thanks @noguchi-takahiro. Further confirmation from @anirudhupadhyaya: 
<img width="595" alt="image" src="https://github.com/user-attachments/assets/359f2782-716b-4e60-a645-9201331a8649">

elsevers committed:
add simulink model of pll

noguchi-takahiro committed:
Add pll test files

noguchi-takahiro commented:
@elsevers I have temporarily added these [pll_test.m](https://github.com/Severson-Group/docs.amdc.dev/blob/1086a1366699e1d1e38d4b227d390c165fb1b65f/source/getting-started/control-with-amdc/encoder-fb/resources/pll_test.m) and [pll_encoder.slx](https://github.com/Severson-Group/docs.amdc.dev/blob/1086a1366699e1d1e38d4b227d390c165fb1b65f/source/getting-started/control-with-amdc/encoder-fb/resources/pll_encoder.slx) to this branch. To simplify the process to create FRF, I made minor adjustments for me to the blocks for continuous-time domain. If you run [pll_test.m](https://github.com/Severson-Group/docs.amdc.dev/blob/1086a1366699e1d1e38d4b227d390c165fb1b65f/source/getting-started/control-with-amdc/encoder-fb/resources/pll_test.m), the following plot is generated:

<p align="center">
<img width="600" alt="image" src="https://github.com/user-attachments/assets/01746aff-6607-44c6-bdc4-063b8665a46c">
</p>

The blue circles in the plot represents the FRF of $\theta_{\text{out}}/\theta_{\text{in}}$ by injecting the chirp signal to $\theta_{\text{in}}$ of the PLL. From my understanding, this PLL is consist of PI and an integrator. The closed-loop (CL) transfer function is as follows:

$$
\frac{\theta_{\text{out}}}{\theta_{\text{in}}} = \frac{K_p s + K_i}{s^2 + K_p s + K_i}
$$

The red line in the plot (PLL) represents this function, which seems to behave like a filter. 

Based on this block [pll_encoder.slx](https://github.com/Severson-Group/docs.amdc.dev/blob/1086a1366699e1d1e38d4b227d390c165fb1b65f/source/getting-started/control-with-amdc/encoder-fb/resources/pll_encoder.slx), PI gains are determined by $K_p = -2\pi \text{Pole1}- 2\pi \text{Pole2}=-\omega_1-\omega_2$ and $K_i = 2\pi \text{Pole1} \times 2\pi \text{Pole2}=\omega_1\omega_2$, where Pole1 = -10 Hz and Pole2 = -100 Hz. Therefore,

$$
\frac{\theta_{\text{out}}}{\theta_{\text{in}}} = \frac{K_p s + K_i}{s^2 + K_p s + K_i} = \frac{(-\omega_1-\omega_2) s + \omega_1\omega_2}{s^2 + (-\omega_1-\omega_2) s +  \omega_1\omega_2} = \frac{(-\omega_1-\omega_2) s + \omega_1\omega_2}{(s-\omega_1)(s-\omega_2)}
$$

This indicates for sure these Pole1 and Pole2 are poles of CL. But I am still thinking about the relationship between the desired pole locations and the roll off frequency of the filter. I am not sure if this is correct approaches but assuming the natural frequency and damping of this CL:

$$
\omega_n = \sqrt{K_i} = \sqrt{\omega_1 \omega_2} = 198 (\text{rad/s}) = 31 (\text{Hz})
$$

$$
\zeta = \frac{K_p}{2\sqrt{K_i}} = \frac{-\omega_1-\omega_2}{2\sqrt{\omega_1\omega_2}} = 1.74
$$

probably the poles are actually determined based on these parameters?

noguchi-takahiro commented:
@elsevers @Daehoon-Sung 

> I want to understand whether this matches what we are doing in BP4 (can you be ready to show me how this is handled in BP4 in Simulink and C code?)

### Simulink
@Daehoon-Sung could you bring your laptop and be ready to open this ([controller_4DOF.slx](https://github.com/Severson-Group/ARL-eturbo/blob/main/BP4/Control/controller_4DOF.slx) that I think `main` branch is okay).

![image](https://github.com/user-attachments/assets/8ef1a261-9a71-4138-8df5-0f6ab74040dd)


### C-code
Here is the C-code of how to implement encoder processing in BP4:

https://github.com/Severson-Group/ARL-eturbo/blob/2888cb5d85e84f6e79ff8f1de7a4df0e4bb8ddbe/Embedded/My-C-Code/usr/bm_4dof/task_bm_4dof.c#L139-L160

The previous code had the issue discussed from [here](https://github.com/Severson-Group/docs.amdc.dev/pull/103#issuecomment-2453031626), but we fixed it via [this PR](https://github.com/Severson-Group/ARL-eturbo/pull/572).



> I am also not sure that it makes sense to have a block diagram like Anirudh is showing and would like your opinion on this.

I think showing the block diagram is good idea still, but we should modify it like the following image:

![image](https://github.com/user-attachments/assets/04ddd326-6fc4-4623-9e2b-393a23a359dd)

I guess this was originally added based on [this comment](https://github.com/Severson-Group/docs.amdc.dev/pull/103#issuecomment-2185266907), where we wanted to add "master" block diagram. We can think about this. Also, we should make this more nicely using Inkscape [like this](https://docs.amdc.dev/hardware/subsystems/encoder.html#block-diagram).

### Useful reference
See this [PR,](https://github.com/Severson-Group/ARL-eturbo/pull/434) especially [this comment](https://github.com/Severson-Group/ARL-eturbo/pull/434#issuecomment-2452805793) is useful too.

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,159 @@
+
+As a first step, the user may use the AMDC `drv/encoder` driver module function `encoder_get_position()` to get the count of the encoder reading. The`drv/encoder` driver module also has a function called `encoder_get_steps()` which gives the incremental change in the encoder position. Whereas, `encoder_get_position()` gives the actual position of the shaft and this can be converted to rotor posit ...
+
+ Next, the user needs to verify if the encoder count is increasing or decreasing with counter-clock wise rotation of shaft. This may be done manually by rotating the shaft and observing the trend of the reported position with respect to the direction of rotation. In the figure, the signal `CCW` indicates this directionality. It must be  set to `1` if the encoder count increases with counter clock ...
```
For the variable `CCW` here, I personally prefer to use more clear variable name. For example, `fis_counter_clockwise` or `ccw_rotation_flag`. Or maybe we should come up with nicer one.

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -1,13 +1,159 @@
+
+
+
+### Finding the offset
```
@Daehoon-Sung I am sure you know well about this section and contribute to update some.

elsevers commented:
- Indicate a complete revolution is needed before using the encoder to find the z pulse
- @elsevers will look into discrepancy in counts vs pulses per revolution (note that BP4 uses this encoder `RM44IC0010B10F2E96`, which the datasheet says has CPR of 1024, however the firmware labels that as PPR and uses it both to initialize the PPR value and to convert counts to an angle)
- @Daehoon-Sung to create a branch and PR revising the offset calculation
- @noguchi-takahiro to come up with a plan for how to revise the observer section as well as including/presenting the PLL block

Daehoon-Sung committed:
Change indexme file

Daehoon-Sung committed:
Update index.md

elsevers commented:
I have confirmed that the AMDC firmware and fpga logic is using `pulses per revolution` (PPR) to mean `counts per revolution` (CPR). This does not match the definition laid out in [the AMDC's encoder subsystem documentation](https://docs.amdc.dev/hardware/subsystems/encoder.html), where CPR is based quadrature encoding and therefore we should have `CPR = 4xPPR`.

This is confusing and I think we should change our terminology in the firmware. Hmm. That will be a separate issue.

elsevers committed:
Edit finding offset and step 1

elsevers committed:
Edit finding the offset section

elsevers committed:
Add encoder angle diagram, first edits of text to use it.

elsevers committed:
Add latex source for motor cross section

elsevers committed:
add test code

elsevers committed:
Merge branch 'add-control-content' into add-enc-anirudh

elsevers committed:
- Edit everything before finding the offset

- Fix links

elsevers committed:
Merge branch 'add-enc-anirudh' into user/Daehoon-Sung/update-offset-report

elsevers commented:
@Daehoon-Sung and @noguchi-takahiro,

A few things to be aware of:

1. I pushed up key change that include adding this figure and using it throughout the document in the descriptions / code changes: <img width="200" alt="image" src="https://github.com/user-attachments/assets/a29d84f5-bb16-4566-9d42-d629faa3557e" />
2. I am updating the writing to assume that we will edit the encoder driver in the AMDC to change the functions to refer to counts instead of pulses, i.e. ~`encoder_set_pulses_per_rev()`~ -> `encoder_set_counts_per_rev()`
3. I merged the new changes into both of your branches.

elsevers committed:
change image naming to dash-case

elsevers committed:
Improve typesetting around the image

elsevers committed:
Add torque characteristic and update step 1 instructions.

elsevers committed:
Make motor cross-section plot 2 poles, add phave v and w axes

elsevers committed:
Merge branch 'add-enc-anirudh' into user/Daehoon-Sung/update-offset-report

elsevers committed:
Add clarifying comment to torque characteristic

Daehoon-Sung committed:
Update index.md

Apply feedback from Professor's comment (option 2) to revise the steps to find offset

Daehoon-Sung committed:
Update the encoder offset plot

Daehoon-Sung committed:
Update index.md

Update the image

Daehoon-Sung committed:
Update index.md

Add the formula

Daehoon-Sung committed:
Update source/getting-started/control-with-amdc/encoder-fb/index.md

Co-authored-by: Takahiro <114006024+noguchi-takahiro@users.noreply.github.com>

Daehoon-Sung committed:
Update index.md

Daehoon-Sung committed:
Update index.md

Daehoon-Sung committed:
Update index.md

noguchi-takahiro committed:
Fix rendering of equation

noguchi-takahiro committed:
Fix rendering issue of image

Daehoon-Sung committed:
Update the encoder image

Daehoon-Sung committed:
Merge branch 'user/Daehoon-Sung/update-offset-report' of https://github.com/Severson-Group/docs.amdc.dev into user/Daehoon-Sung/update-offset-report

Daehoon-Sung committed:
Update index.md

Update the variable names

Daehoon-Sung committed:
Update the figure

Daehoon-Sung committed:
Clarify terminology for mechanical angle in documentation

Daehoon-Sung committed:
Enhance encoder offset determination section

Added detailed explanation and equations for determining encoder offset using closed-loop control and voltage measurements.

Daehoon-Sung committed:
Clarify voltage equation with angular velocity note

Added clarification about electrical angular velocity in the voltage equation.

Daehoon-Sung committed:
Clarify voltage vector description in index.md

Daehoon-Sung committed:
Fix voltage vector equation formatting

Corrected the formatting of the voltage vector equation in complex vector form.

Daehoon-Sung committed:
Refine description of encoder offset estimation process

Clarified the conditions for estimating encoder offset by refining language regarding the alignment of the gamma-delta and d-q frames.

Daehoon-Sung committed:
Fix minor grammatical error in encoder offset section

Daehoon-Sung committed:
Fix formatting in encoder offset determination section

Daehoon-Sung committed:
Revise current command notation in encoder equations

Updated notation for current commands in the equation for v_gamma.

noguchi-takahiro committed:
Fix rendering issues for images

Daehoon-Sung committed:
Clarify angle calculation and encoder offset procedure

Updated explanations regarding the calculation of the motor's angle and the determination of the encoder offset. Improved clarity on the relationship between electrical angle and voltage vector.

elsevers committed:
Revise encoder offset section contents (#146)

* Edit encoder offset precise section

* Improve typesetting and clarify vd

elsevers committed:
Merge pull request #139 from Severson-Group/user/Daehoon-Sung/update-offset-report

Update offset section in the encoder article

elsevers commented:
@mhmokhtarabadi here is the most recent build of this article:
[Encoder Feedback - AMDC Platform.pdf](https://github.com/user-attachments/files/26471581/Encoder.Feedback.-.AMDC.Platform.pdf)

elsevers committed:
fix typo

elsevers committed:
clean up reference

elsevers committed:
cleanup msc lint errors

elsevers committed:
fix reference warning and eq typsetting

elsevers committed:
rename files to dash case

elsevers committed:
fix typo

elsevers committed:
replace encoder block diagram

elsevers reviewed (approved):
Let's move forward with this.
